File: data/src/main/kotlin/dev/yashgarg/kimai/api/KimaiRepository.kt
/*
 * Copyright © 2023 Yash Garg.
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */
package dev.yashgarg.kimai.api

import dev.yashgarg.kimai.models.Activity
import dev.yashgarg.kimai.models.Ping
import dev.yashgarg.kimai.models.TimesheetActivity
import dev.yashgarg.kimai.util.ApiResult

interface KimaiRepository {
  suspend fun ping(): ApiResult<Ping>

  suspend fun getActivities(
    projectId: Int? = null,
    projects: List<Int> = emptyList(),
    visible: Int = 1,
    globals: Boolean? = null,
    orderBy: String? = null,
    order: String = "ASC",
    searchTerm: String? = null,
  ): ApiResult<List<Activity>>

  suspend fun getProjects(
    visible: Int = 1,
  ): ApiResult<List<Activity>>

  suspend fun getTimeSheets(
    orderBy: String? = null,
    order: String? = null,
  ): ApiResult<List<TimesheetActivity>>

  suspend fun createTimeSheet(
    begin: String,
    end: String? = null,
    description: String? = null,
    tags: List<String>? = null,
    project: Int? = null,
    activity: Int? = null,
  ): ApiResult<TimesheetActivity>
}

File: lib/data/api/kimai_repository.dart
import 'package:fpdart/fpdart.dart';
import 'package:kimai/data/models/auth_user.dart';

import 'package:kimai/data/models/models.dart';

abstract class KimaiRepository {
  Future<Either<Ping, Exception>> ping(AuthUser user);

  Future<Either<List<Activity>, Exception>> getActivities({
    int? projectId,
    List<int>? projects,
    int visible = 1,
    bool? globals,
    String? orderBy,
    String order = 'ASC',
    String? searchTerm,
  });

  Future<Either<List<Activity>, Exception>> getProjects({
    int visible = 1,
  });

  Future<Either<List<TimesheetActivity>, Exception>> getTimeSheets({
    String? orderBy,
    String? order,
  });

  Future<Either<List<Customer>, Exception>> getCustomers({
    int visible = 1,
  });

  Future<Either<TimesheetActivity, Exception>> createTimeSheet({
    required String begin,
    String? end,
    required int project,
    required int activity,
  });

  Future<Either<UserEntity, Exception>> getMyUser();
}
